README.md
# i2p.plugins.tor-updater

A Tor package updater and runner as an I2P Plugin. This plugin is
usable on Windows, Linux, and OSX, as is the freestanding binary.
This also functions as a freestanding update for the Tor Browser
Bundle and is capable of configuring Tor Browser from the terminal
and updating it without running it, should the user choose to operate
this way. This is now more like a fully-fledged browser management
tool that also manages a long-running Tor process if it needs it.

This project explicitly views I2P and Tor as complementary privacy
tools and explores good ways of combining them in a user-facing tool.
It is especially useful when combined with a SOCKS Outproxy plugin
like the one written by zzz. You can obtain it within I2P: http://zzz.i2p/topics/3219

Usage:
------

See [Usage](USAGE.md) for command-line usage. When used from the Terminal,
it acts as a cross-platform Tor Browser installer and wrapper, which adds
non-Tor specific options.

[HTML version](usage.html)
Plugin:
-------

Clearnet visitor? You'll need to use the [Github Releases Mirror](https://github.com/eyedeekay/i2p.plugins.tor-manager/releases/).

- [i2p.plugins.tor-manager-linux-386](https://github.com/eyedeekay/i2p.plugins.tor-manager/releases/download/0.0.12/i2p.plugins.tor-manager-linux-386.su3)
- [i2p.plugins.tor-manager-windows-amd64](https://github.com/eyedeekay/i2p.plugins.tor-manager/releases/download/0.0.12/i2p.plugins.tor-manager-windows-amd64.su3)
- [i2p.plugins.tor-manager-darwin-arm64](https://github.com/eyedeekay/i2p.plugins.tor-manager/releases/download/0.0.12/i2p.plugins.tor-manager-darwin-arm64.su3)
- [i2p.plugins.tor-manager-linux-amd64](https://github.com/eyedeekay/i2p.plugins.tor-manager/releases/download/0.0.12/i2p.plugins.tor-manager-linux-amd64.su3)
- [i2p.plugins.tor-manager-windows-386](https://github.com/eyedeekay/i2p.plugins.tor-manager/releases/download/0.0.12/i2p.plugins.tor-manager-windows-386.su3)
- [i2p.plugins.tor-manager-darwin-amd64](https://github.com/eyedeekay/i2p.plugins.tor-manager/releases/download/0.0.12/i2p.plugins.tor-manager-darwin-amd64.su3)

Status:
-------

![Screenshot 2](screenshot-console.png)

Linux: Usable, everything implemented works.
Windows: Usable, everything implemented works.
OSX: Usable, everything implemented works.

Other systems are not targeted and should use a Tor binary built from source,
provided by TPO or, their prefered package management system and not this plugin.
The plugin will not start a Tor instance if a SOCKS proxy is open on port 9050.

![Screenshot](screenshot-i2pbrowser.png)

### Primary Goals(Done):

1. Ship known-good public keys, download a current Tor for the platform in the background, authenticate it, and launch it only if necessary.
 - Works on Windows, Linux, OSX
2. Supervise Tor as a ShellService plugin to I2P
 - Works on Linux, Windows, OSX
3. Keep Tor up-to-date
 - Works on Windows, Linux, OSX
4. Work as an I2P Plugin OR as a freestanding app to be compatible with all I2P distributions
 - Works on Linux, Windows, OSX
5. Download Tor Browser from an in-I2P mirror(or one of a network of in-I2P mirrors)
 - Works on Linux, Windows, OSX

### Secondary Goals(Also done):

1. Launch Tor Browser
 - Works on Linux, Windows, OSX
2. Configure and launch Tor browser for use with I2P
 - Works on Linux, Windows, OSX

#### Optional Features I might add if there is interest(Not much stopping me...):

1. Mirror the files which it downloads to an I2P Site
 - Works on Windows, Linux, OSX
2. Mirror the files which it downloads to I2P torrents
 - Works on Windows and Linux, might work on OSX but unsure.
3. Set up an onion site which announces an I2P mirror exists
 - Works on Windows, Linux, OSX
4. Use Bittorrent-over-I2P to download the Tor Browser software
 - Worksish. Still a little janky. Usable on any platform if you're a little patient.
5. Embed jpackaged I2P routers and manage them internally
 - Works on Windows and Linux. Can be done on OSX but needs to be different. AMD64 only.
6. Encrypt the "Working directory" with all the plugin data using a password.
 - Works on Windows, Linux, and OSX.
7. Use a transparently encrypted filesystem to interact with the host system.
 - Not done.
8. Option to use BRB in a thread as an in-I2P replacement for `mibbit` IRC client.
 - Works on Windows, Linux, OSX

### Usage as a Library

[More information at the GoDoc](https://pkg.go.dev/i2pgit.org/idk/i2p.plugins.tor-manager)

This is also useful as a library for downloading a Tor Browser Bundle. This API
isn't really stable, more "stabilizing." Feel free to use it, but it may still
change a little.

To create a new instance, use:

``` Go
client, err = tbserve.NewClient(*verbose, *lang, *system, *arch, &content)
```

Customize the client using the exposed variables and methods:

``` Go
client.Host = *host
client.Port = *port
client.TBS.Profile = &content
client.TBS.PassThroughArgs = flag.Args()
```

And serve the controller:

``` Go
if err := client.Serve(); err != nil {
  log.Fatal(err)
}
```

### Similar Projects:

- [https://github.com/micahflee/torbrowser-launcher](https://github.com/micahflee/torbrowser-launcher)
- [https://github.com/whonix/tb-updater](https://github.com/whonix/tb-updater)
- Probably about 200 or so bespoke scripts written and abandoned by hackers over the years, search github

#### Why this is Better than those projects(A.K.A Why didn't you just use...)

##### [TorBrowser Launcher](https://github.com/micahfleee/torbrowser-launcher)

i2p.plugins.tor-manager is in many ways a clone of Tor Browser Launcher by Micah Lee,
it reproduces all of TBL's functionality for downloading, verifying, and running
Tor Browser. In particular Micah's project deserves credit for the AppArmor profiles,
which were used in the apparmor profile generator functionality and will probably
continue to be the upstream example which I use in my code. Micah's also already got
a PPA, a FlatPak, a Snap, an `.rpm` and a bunch of other stuff that provides, through
packaging, a better way to deal with updates to his software and a better way to deal
with the supply chain for many users. There is no reason that `i2p.plugins.tor-manager`
cannot do this, in fact it's easier for us than it is for Micah, but this project is
about a month and a half old and his is several years old. I haven't done a bunch of
packages yet.

Improvements on Tor Browser Launcher include:

 - Does not require GPG to be in the `$PATH` or on the system at all
 - Capable of managing a long-running Tor service without also running the Tor Browser
 - Works everywhere Tor Browser does, not just Linux. Including:
  Linux, Mac, Windows, AMD64, X86, and M1
 - Embeds all required resources into a single binary which unpacks itself to a directory
  of the user's choice. Never touches any other directory(in ~/.cache or ~/.local, etc),
  unless specifically instructed to.
 - Easy to package by placing a single static binary into a boilerplate package.
 - Automatically generate an AppArmor profile which works for custom installation directory
 - Accepts a custom profile argument without interfering with normal Tor Browser
  functionality.
 - Bundles a "Extra" profiles for I2P browsing, non-Anonymous browsing, and
  offline/localhost-only browsing.
 - Automatically organize multiple Tor Browser profiles for different networks(I2P,
  Clearnet) and discourage accidental mixing.
 - Optional TAILS mode, adds uBlock Origin to Tor Browser
 - Embeds an I2P router in case the host system doesn't have an I2P router
 - Encrypt the working directory automatically when it's not in use.
 - No python. Language wars are stupid, but so is python.

Things Tor Browser Launcher does that this does not(yet):

 - Micah went to far greater effort to assure that TPO's HTTPS certificate was
 authentic and avoid the Certificate Authority/PKI system. As a result, CA's are
 untrusted entities when downloading from the default Tor Project site. This is
 probably a pretty good idea.
 - This application does not offer a choice of downloading Tor Browser over the
 clearnet if there is literally any other option available to it. This is roughly equivalent
 to downloading Tor Browser over system Tor being always on in Tor Browser Launcher.
 - This application does not play dial-up noises.

#### [TB-Updater(Whonix)](https://github.com/whonix/tb-updater)

tb-updater is generally only useful on Linux systems and especially Debian-based systems,
and only recieves serious testing within the Whonix system itself. It's a highly specialized
tool in some ways, depending on packages which must be installed and repositories which must
be configured. For their purposes, this is probably a good thing, Whonix's raison d'etre is
to carefully configure every aspect of their system to be as anonymous and secure as possible,
and their approach deals with supply-chain issues which `i2p.plugins.tor-manager` can only
deal with in **packaging** and not by adding features to the launcher itself. Therefore, while
most of the improvements on Tor Browser Launcher also apply to TB-Updater, the
`i2p.plugins.tor-manager` and `tb-launcher` are actually less directly comparable.

<a href="https://www.flaticon.com/free-icons/garlic" title="garlic icons">Garlic icons created by Icongeek26 - Flaticon</a>
<a href="https://www.flaticon.com/free-icons/onion" title="onion icons">Onion icons created by Freepik - Flaticon</a>
<a href="https://www.flaticon.com/free-icons/search" title="search icons">Search icons created by Freepik - Flaticon</a>
<a href="https://www.flaticon.com/free-icons/offline" title="offline icons">Offline icons created by Flat Icons - Flaticon</a>

### More Screenshots:

- ![Screenshot](screenshot.png)
- ![Screenshot](screenshot-dark.png)


debian/changelog
i2p.plugins.tor-manager (0.0.13-1) UNRELEASED; urgency=medium

  * tag 0.0.13

 -- idk <hankhill19580@gmail.com> Thu, 07 Jul 2022 11:24:17 -0400


i2p.plugins.tor-manager (0.0.12-1) UNRELEASED; urgency=medium

  * tag 0.0.12
  * 239f269 add resource attribution to LICENSE file
  * cf6eaf8 add flatpak desktop files
  * 6640575 add flatpak desktop files
  * 831911e work on maintainint 2 TBB copies
  * cb8730a work on maintainint 2 TBB copies
  * 259f110 work on maintainint 2 TBB copies
  * 5c48249 exit after mirroring everything
  * 7a6e73b improve mirroring functionality
  * 8970a1f improve mirroring functionality
  * 15a5b45 improve mirroring functionality
  * 450512f improve mirroring functionality
  * 6d0196b improve mirroring functionality
  * f6478b9 don't create so many Tor clients, build mirroring into the system, download to subdir of i2psnark dir
  * 8228621 update the modules again
  * f8d890f unset transport before and after re-setting transport
  * d2a035d add hardened firefox, offline mode to menu
  * 63b7289 add hardened firefox, offline mode to menu

 -- idk <hankhill19580@gmail.com> Thu, 07 Jul 2022 11:24:17 -0400


i2p.plugins.tor-manager (0.0.10-1) UNRELEASED; urgency=medium

  * tag 0.0.10
  * 239f269 add resource attribution to LICENSE file
  * cf6eaf8 add flatpak desktop files
  * 6640575 add flatpak desktop files
  * 831911e work on maintainint 2 TBB copies
  * cb8730a work on maintainint 2 TBB copies
  * 259f110 work on maintainint 2 TBB copies
  * 5c48249 exit after mirroring everything
  * 7a6e73b improve mirroring functionality
  * 8970a1f improve mirroring functionality
  * 15a5b45 improve mirroring functionality
  * 450512f improve mirroring functionality
  * 6d0196b improve mirroring functionality
  * f6478b9 don't create so many Tor clients, build mirroring into the system, download to subdir of i2psnark dir
  * 8228621 update the modules again
  * f8d890f unset transport before and after re-setting transport
  * d2a035d add hardened firefox, offline mode to menu
  * 63b7289 add hardened firefox, offline mode to menu
  * 943de35 clarify why I2P mirror is not in use yet
  * 1fe014e try and get the systray tto work on OSX
  * 6deedad switch to wget-ds to download su3 files
  * dcba130 comment out shellservice copy line it's not required anymore
  * b5f30e6 add compatibility mode for NSIS args
  * 06bb27f start another manifesto

 -- idk <hankhill19580@gmail.com> Thu, 07 Jul 2022 11:24:17 -0400


i2p.plugins.tor-manager (0.0.8-1) UNRELEASED; urgency=medium

  * tag 0.0.8
  * 239f269 add resource attribution to LICENSE file
  * cf6eaf8 add flatpak desktop files
  * 6640575 add flatpak desktop files
  * 831911e work on maintainint 2 TBB copies
  * cb8730a work on maintainint 2 TBB copies
  * 259f110 work on maintainint 2 TBB copies
  * 5c48249 exit after mirroring everything
  * 7a6e73b improve mirroring functionality
  * 8970a1f improve mirroring functionality
  * 15a5b45 improve mirroring functionality
  * 450512f improve mirroring functionality
  * 6d0196b improve mirroring functionality
  * f6478b9 don't create so many Tor clients, build mirroring into the system, download to subdir of i2psnark dir
  * 8228621 update the modules again
  * f8d890f unset transport before and after re-setting transport
  * d2a035d add hardened firefox, offline mode to menu
  * 63b7289 add hardened firefox, offline mode to menu
  * 943de35 clarify why I2P mirror is not in use yet
  * 1fe014e try and get the systray tto work on OSX
  * 6deedad switch to wget-ds to download su3 files
  * dcba130 comment out shellservice copy line it's not required anymore
  * b5f30e6 add compatibility mode for NSIS args
  * 06bb27f start another manifesto
  * 1210b6c More torrent subsystem upgrades
  * 4057c96 More torrent subsystem upgrades
  * e6a46e8 More torrent subsystem upgrades
  * 589f825 More torrent subsystem upgrades
  * 20e21b5 More torrent subsystem upgrades
  * 3ce1c7d More torrent subsystem upgrades
  * 58a23a2 Make torrent updates a little better
  * 35085d3 unset default lang in makefile
  * 7342f67 unset default lang in makefile
  * 7a36c48 Fix torrent downloaded size check. More-or-less..
  * 9ce7c2d Add ability to get the Tor version without instantiating a whole object.
  * 89319b9 Start working on containerized 32-bit builds
  * f122740 move I2P launcher to own file

 -- idk <hankhill19580@gmail.com> Thu, 07 Jul 2022 11:24:18 -0400


i2p.plugins.tor-manager (0.0.7-1) UNRELEASED; urgency=medium

  * tag 0.0.7
  * 239f269 add resource attribution to LICENSE file
  * cf6eaf8 add flatpak desktop files
  * 6640575 add flatpak desktop files
  * 831911e work on maintainint 2 TBB copies
  * cb8730a work on maintainint 2 TBB copies
  * 259f110 work on maintainint 2 TBB copies
  * 5c48249 exit after mirroring everything
  * 7a6e73b improve mirroring functionality
  * 8970a1f improve mirroring functionality
  * 15a5b45 improve mirroring functionality
  * 450512f improve mirroring functionality
  * 6d0196b improve mirroring functionality
  * f6478b9 don't create so many Tor clients, build mirroring into the system, download to subdir of i2psnark dir
  * 8228621 update the modules again
  * f8d890f unset transport before and after re-setting transport
  * d2a035d add hardened firefox, offline mode to menu
  * 63b7289 add hardened firefox, offline mode to menu
  * 943de35 clarify why I2P mirror is not in use yet
  * 1fe014e try and get the systray tto work on OSX
  * 6deedad switch to wget-ds to download su3 files
  * dcba130 comment out shellservice copy line it's not required anymore
  * b5f30e6 add compatibility mode for NSIS args
  * 06bb27f start another manifesto
  * 1210b6c More torrent subsystem upgrades
  * 4057c96 More torrent subsystem upgrades
  * e6a46e8 More torrent subsystem upgrades
  * 589f825 More torrent subsystem upgrades
  * 20e21b5 More torrent subsystem upgrades
  * 3ce1c7d More torrent subsystem upgrades
  * 58a23a2 Make torrent updates a little better

main.go
package main

import (
	"embed"
	"fmt"
	"io/ioutil"
	"log"
	"net"
	"os"
	"os/signal"
	"path/filepath"
	"runtime"
	"strings"

	flag "github.com/spf13/pflag"

	"github.com/cloudfoundry/jibber_jabber"
	"github.com/eyedeekay/go-I2P-jpackage"
	"github.com/itchio/damage"
	"github.com/itchio/damage/hdiutil"
	"github.com/itchio/headway/state"
	"github.com/ncruces/zenity"
	tbget "i2pgit.org/idk/i2p.plugins.tor-manager/get"
	tbserve "i2pgit.org/idk/i2p.plugins.tor-manager/serve"
	tbsupervise "i2pgit.org/idk/i2p.plugins.tor-manager/supervise"

	tinymce "github.com/eyedeekay/go-htmleditor"
)

/*
TODO: A "Default" config file which uses hardened Tor Browser for clearnet
(or default-route) browsing.
*/

//go:embed tor-browser/unpack/i2p.firefox/*
//go:embed tor-browser/unpack/i2p.firefox.config/*
//go:embed tor-browser/unpack/awo@eyedeekay.github.io.xpi
//go:embed tor-browser/TPO-signing-key.pub
//go:embed tor-browser/NOT-TPO-signing-key.pub
//go:embed garliconion.png
//go:embed onion.png
//go:embed www.png
//go:embed offline.png
//go:embed torbrowser.desktop
//go:embed i2ptorbrowser.desktop
//go:embed LICENSE
var content embed.FS

// print license, then exit 0
func LICENSE() {
	//get LICENSE from embedded content
	license, err := content.Open("LICENSE")
	if err != nil {
		log.Fatal(err)
	}
	license_bytes, err := ioutil.ReadAll(license)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println(string(license_bytes))
	I2P.PrintLicenses()
	os.Exit(0)
}

func OS() string {
	switch runtime.GOOS {
	case "darwin":
		return "osx"
	case "linux":
		return "linux"
	case "windows":
		return "win"
	default:
		return "unknown"
	}
}

func ARCH() string {
	switch runtime.GOARCH {
	case "386":
		return "32"
	case "amd64":
		return "64"
	case "arm64":
		if OS() == "osx" {
			return "64"
		}
		return ""
	default:
		return "unknown"
	}
}

var theLang = os.Getenv("TBLANG")

func defaultLang() string {
	if theLang != "" {
		return theLang
	}
	lang, _ := jibber_jabber.DetectIETF()
	return lang
}

func defaultTor() bool {
	_, err := net.Listen("TCP", "127.0.0.1:9050")
	if err != nil {
		return true
	}
	return false
}

var (
	lang       = flag.String("lang", defaultLang(), "Language to download")
	system     = flag.String("os", OS(), "OS/arch to download")
	arch       = flag.String("arch", ARCH(), "OS/arch to download")
	i2pbrowser = flag.Bool("i2pbrowser", false, "Open I2P in Tor Browser")
	i2pconfig  = flag.Bool("i2pconfig", false, "Open I2P routerconsole in Tor Browser with javscript enabled and non-routerconsole sites disabled")
	torbrowser = flag.Bool("torbrowser", false, "Open Tor Browser")
	i2peditor  = flag.Bool("i2peditor", false, "Open I2P Site Editor in Tor Browser")
	verbose    = flag.Bool("verbose", false, "Verbose output")
	directory  = flag.String("directory", DefaultDir(), "Directory operate in")
	host       = flag.String("host", "127.0.0.1", "Host to serve on")
	port       = flag.Int("port", 7695, "Port to serve on")
	bemirror   = flag.Bool("bemirror", false, "Act as an in-I2P mirror when you're done downloading")
	shortcuts  = flag.Bool("shortcuts", false, "Create desktop shortcuts")
	apparmor   = flag.Bool("apparmor", false, "Generate apparmor rules")
	offline    = flag.Bool("offline", false, "Work offline. Differs from Firefox's offline mode in that cannot be disabled until the browser is closed.")
	clearnet   = flag.Bool("clearnet", Clearnet(), "Use clearnet (no Tor or I2P) in Tor Browser")
	profile    = flag.String("profile", "", "use a custom profile path, normally blank")
	help       = flag.Bool("help", false, "Print help and quit")
	mirror     = flag.String("mirror", Mirror(), "Mirror to use. I2P will be used if an I2P proxy is present, if system Tor is available, it will be downloaded over the Tor proxy.")
	solidarity = flag.Bool("onion", defaultTor(), "Serve an onion site which shows some I2P propaganda")
	torrent    = flag.Bool("torrent", tbget.TorrentReady(), "Create a torrent of the downloaded files and seed it over I2P using an Open Tracker")
	destruct   = flag.Bool("destruct", false, "Destructively delete the working directory when finished")
	password   = flag.String("password", Password(), "Password to encrypt the working directory with. Implies -destruct, only the encrypted container will be saved.")
	chat       = flag.Bool("chat", false, "Open a WebChat client")
	notor      = flag.Bool("notor", false, "Do not automatically start Tor")
	nounpack   = flag.Bool("nounpack", false, "Do not unpack the Tor Browser")
	ptop       = flag.Bool("p2p", tbget.TorrentDownloaded(defaultLang(), OS()+ARCH()), "Use bittorrent over I2P to download the initial copy of Tor Browser")
	torversion = flag.Bool("torversion", false, "Print the version of Tor Browser that will be downloaded and exit")
	mirrorall  = flag.Bool("mirrorall", false, "Download and mirror every language and OS/arch combination")
	nevertor   = flag.Bool("nevertor", false, "Never use Tor for downloading Tor Browser")
	license    = flag.Bool("license", false, "Print the license and exit")
)

func Clearnet() bool {
	if tmc := os.Getenv("TOR_MANAGER_CLEARNET"); tmc != "" {
		switch tmc {
		case "1", "true", "yes", "on":
			return true
		}
	}
	return false
}

func Password() string {
	require_password := os.Getenv("TOR_MANAGER_REQUIRE_PASSWORD")
	if require_password == "" && !PluginStat() {
		require_password = "true"
	}
	// search os.Args for -license flag and if it's present, set to false
	for _, arg := range os.Args {
		if arg == "-license" {
			require_password = "false"
		}
	}

	switch require_password {
	case "true", "1", "yes", "on":
		passwd, err := zenity.Entry(
			"Enter a password if you want to encrypt the working directory",
			zenity.Title("Work Directory Encryption"),
			zenity.CancelLabel("Don't encrypt"),
			zenity.OKLabel("Encrypt"),
			zenity.Width(400),
			zenity.EntryText("password"),
		)
		if err != nil {
			if !strings.Contains(err.Error(), "canceled") {
				log.Panicln(err)
			}
			log.Println("Password dialog canceled")
			return ""
		}
		return passwd
	}
	return ""
}

func Mirror() string {
	if mir := os.Getenv("TOR_MANAGER_MIRROR"); mir != "" {
		fmt.Println("Using environment mirror", mir)
		return mir
	}
	if runtime.GOOS == "linux" && runtime.GOARCH == "arm64" {
		fmt.Println("Using arm64 mirror")
		return "https://sourceforge.net/projects/tor-browser-ports/files"
	}
	clear := os.Getenv("TOR_MANAGER_CLEARNET")
	switch clear {
	case "1", "true", "yes", "on":
		fmt.Println("Using clearnet mirror")
		return "https://dist.torproject.org/torbrowser/"
	}
	clearmirror := os.Getenv("TOR_MANAGER_CLEARNET_MIRROR")
	switch clearmirror {
	case "1", "true", "yes", "on":
		fmt.Println("Using clearnet mirror")
		return "https://dist.torproject.org/torbrowser/"
	}
	if tbget.Torrent(*lang, OS()+ARCH()) {
		fmt.Println("Using torrent mirror")
		return "http://localhost:7657/i2psnark/"
	}
	if tbget.TestHTTPDefaultProxy() {
		//fmt.Println("Using I2P mirror")
		//return "http://dist.torproject.i2p/torbrowser/"
		fmt.Println("Using clearnet mirror instead of I2P mirror due to hash sum mismatch issue")
		return "https://dist.torproject.org/torbrowser/"
	}
	fmt.Println("Using clearnet mirror")
	return "https://dist.torproject.org/torbrowser/"
}

var snowflake *bool

var client *tbserve.Client

func main() {
	for _, arg := range os.Args {
		if arg == "-license" {
			LICENSE()
		}
	}
	if err := NSISCompat(); err != nil {
		log.Println("NSIS compat mode failure", err)
		os.Exit(0)
	}
	filename := filepath.Base(os.Args[0])
	SnowflakeFlag()
	usage := flag.Usage
	flag.Usage = func() {
		fmt.Printf("Usage: %s %s\n", filename, "[options]")
		fmt.Printf("\n")
		fmt.Printf("Downloads, verifies and unpacks Tor Browser. Manages the Tor Browser\n")
		fmt.Printf("system in environments where Tor is not in use. Monitors a long-running\n")
		fmt.Printf("Tor process and downloads updates when Tor is not available.\n")
		fmt.Printf("\n")
		fmt.Printf("Options:\n")
		fmt.Printf("\n")
		usage()
		fmt.Printf("\nAvailable Languages:\n\n")
		for _, l := range tbget.Languages() {
			fmt.Printf("  - %s\n", l)
		}
		fmt.Printf("\n")
	}
	args, trailers := CleanupArgs()
	log.Printf("Args: %v\n", args)
	log.Printf("Trailers: %v\n", trailers)
	os.Args = args
	flag.Parse()
	if *nevertor {
		err := os.Setenv("TOR_MANAGER_NEVER_USE_TOR", "true")
		if err != nil {
			log.Panicln(err)
		}
	}
	if *mirrorall {
		err := os.Setenv("TOR_MANAGER_NEVER_USE_TOR", "true")
		if err != nil {
			log.Panicln(err)
		}
		err = mirrorAll()
		if err != nil {
			log.Fatal(err)
		}
		os.Exit(0)
	}
	if *torversion {
		torbrowserversion, err := tbget.GetTorBrowserVersionFromUpdateURL()
		if err != nil {
			log.Panicln(err)
		}
		fmt.Println(torbrowserversion)
		os.Exit(0)
	}
	if *ptop {
		log.Println("Using p2p")
		*mirror = "http://localhost:7657/i2psnark/"
	}
	if *password != "" {
		log.Println("Looking for directory with password")
		DecryptTarXZifThere(*directory, *password)
		// capture sigint
		c := make(chan os.Signal, 1)
		signal.Notify(c, os.Interrupt)
		go func() {
			for range c {
				log.Println("Caught interrupt, exiting")
				err := EncryptTarXZip(*directory, *password)
				if err != nil {
					log.Println(err)
				}
				os.Exit(0)
			}
		}()
	}
	if *clearnet {
		*mirror = "http://dist.torproject.org/torbrowser/"
	}
	if *snowflake {
		go Snowflake()
	}
	if *destruct {
		defer OverwriteDirectoryContents(*directory)
	}
	tbget.WORKING_DIR = *directory
	if filename == "i2pbrowser" {
		log.Println("Starting I2P in Tor Browser")
		*i2pbrowser = true
	} else if filename == "torbrowser" {
		log.Println("Starting Tor Browser")
		*torbrowser = true
	} else if filename == "i2pconfig" {
		log.Println("Starting I2P routerconsole in Tor Browser")
		*i2pconfig = true
	} else if filename == "firefox" || *clearnet || *offline {
		*clearnet = true
	}
	if *profile == "" {
		if *offline {
			*profile = filepath.Join(tbget.WORKING_DIR, "profile.firefox.offline")
		} else if *clearnet {
			*profile = filepath.Join(tbget.WORKING_DIR, "profile.firefox")
		} else {
			*profile = filepath.Join(tbget.WORKING_DIR, "profile.firefox.default")
		}
	} else {
		*profile = filepath.Join(tbget.WORKING_DIR, *profile)
	}
	if *i2pbrowser && *torbrowser {
		log.Fatal("Please don't open I2P and Tor Browser at the same time when running from the terminal.")
	}
	if *lang == "" {
		var err error
		*lang, err = jibber_jabber.DetectIETF()
		if err != nil {
			log.Fatal("Please specify a language", err)
		}
		log.Println("Using auto-detected language", *lang)
	}
	if I2PDaemon, err := StartI2P(*directory); err != nil {
		log.Fatal(err)
	} else {
		if I2PDaemon != nil {
			defer I2PDaemon.Stop()
		}
	}
	var err error
	client, err = tbserve.NewClient(*verbose, *lang, *system, *arch, *mirror, &content, *nounpack)
	if err != nil {
		log.Fatal("Couldn't create client", err)
	}
	if *apparmor {
		err := GenerateAppArmor()
		if err != nil {
			log.Fatal("Couldn't generate apparmor rules", err)
		}
		log.Println("################################################################")
		log.Println("#             AppArmor rules generated successfully            #")
		log.Println("################################################################")
		log.Println("!IMPORTANT! You must now run the following commands:")
		log.Println("sudo mkdir -p /etc/apparmor.d/tunables/")
		log.Println("sudo cp tunables.torbrowser.apparmor /etc/apparmor.d/tunables/torbrowser")
		log.Println("sudo cp torbrowser.Tor.tor.apparmor /etc/apparmor.d/torbrowser.Tor.tor")
		log.Println("sudo cp torbrowser.Browser.firefox.apparmor /etc/apparmor.d/torbrowser.Browser.firefox")
		log.Println("sudo apparmor_parser -r /etc/apparmor.d/tunables/torbrowser")
		log.Println("sudo apparmor_parser -r /etc/apparmor.d/torbrowser.Tor.tor")
		log.Println("sudo apparmor_parser -r /etc/apparmor.d/torbrowser.Browser.firefox")
		log.Println("To copy them to apparmor profiles directory and reload AppArmor")
		return
	}
	if *shortcuts {
		err := CreateShortcuts()
		if err != nil {
			log.Fatal("Couldn't create desktop shortcuts", err)
		}
	}
	client.Host = *host
	client.Port = *port
	client.TBS.Profile = &content
	client.TBS.PassThroughArgs = trailers
	if runtime.GOOS == "darwin" {
		consumer := &state.Consumer{
			OnMessage: func(lvl string, msg string) {
				log.Printf("[%s] %s", lvl, msg)
			},
		}
		host := hdiutil.NewHost(consumer)
		defer damage.Unmount(host, client.TBD.BrowserDir())
	}
	//	log.Fatalf("%s", client.TBS.PassThroughArgs)
	if *help {
		flag.Usage()
		if err := client.TBS.RunTBHelpWithLang(); err != nil {
			log.Fatal(err)
		}
		return
	}
	if *torrent {
		log.Println("Generating I2P torrents of Tor packages")
		if err := client.TBD.GenerateMissingTorrents(); err != nil {
			log.Fatal(err)
		}
		log.Println("I2P torrents generated")
	}
	client.TBS.UnpackI2PAppData()
	client.TBS.UnpackI2PData()
	if *nounpack {
		log.Println("not unpacking, cannot continue")
		os.Exit(0)
	}
	if !(*clearnet || *notor) {
		log.Println("CLEARNET", *clearnet)
		log.Println("NOTOR", *notor)
		client.TBS.RunTorWithLang()
	}

	if *chat {
		log.Println("Starting I2P chat")
		go BRBClient(*directory, "brb")
	}
	go ServeEditor()
	if *i2pbrowser {
		if err := client.TBS.RunI2PBWithLang(); err != nil {
			log.Fatal(err)
		}
	} else if *i2pconfig {
		if err := client.TBS.RunI2PBAppWithLang(); err != nil {
			log.Fatal(err)
		}
	} else if *i2peditor {
		if err := client.TBS.RunI2PSiteEditorWithOfflineClearnetProfile(filepath.Join(client.TBD.UnpackPath, "i2p.firefox.editor")); err != nil {
			log.Fatal(err)
		}
	} else if *torbrowser {
		if err := client.TBS.RunTBWithLang(); err != nil {
			log.Fatal(err)
		}
	} else if *offline {
		log.Println("Working offline")
		if err := client.TBS.RunTBBWithOfflineClearnetProfile(*profile, *offline, *clearnet); err != nil {
			log.Fatal(err)
		}
	} else if *clearnet {
		log.Println("Using a custom profile")
		if err := client.TBS.RunTBBWithOfflineClearnetProfile(*profile, *offline, *clearnet); err != nil {
			log.Fatal(err)
		}
	} else {
		if *bemirror {
			go client.TBD.Serve()
		}
		if *solidarity {
			client.Onion.UnpackSite()
			go ServeOnion()
		}
		go runSysTray(false)
		if err := client.Serve(); err != nil {
			log.Fatal(err)
		}
	}
}

func ServeOnion() error {
	if err := client.Onion.ListenAndServe(); err != nil {
		log.Println("Onion error:", err)
	}
	return nil
}

func ServeEditor() error {
	docroot, err := tbsupervise.FindEepsiteDocroot()
	if err != nil {
		return err
	}
	if err := tinymce.Serve("127.0.0.1", docroot, "index.html", 7685); err != nil {
		log.Println("Couldn't serve editor", err)
	}
	return nil
}

func pathToMe() (string, error) {
	ex, err := os.Executable()
	if err != nil {
		return "", err
	}
	exPath, err := filepath.Abs(ex)
	if err != nil {
		return "", err
	}
	return exPath, nil
}
